### PR TITLE
ci: replace makefile with nox

### PR DESCRIPTION
**Pull request by davidhewitt**
Ports the `Makefile` over to `nox`. Seems like a win for portability!

### CODE REVIEW

**Comment by davidhewitt** [comment]
For simplicity I will probably replace `xtask` with this `noxfile` too

File: noxfile.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+def clippy(session: nox.Session) -> None:
+    for feature_set in ["full", "abi3 full"]:
+        session.run(
+            "cargo",
+            "clippy",
+            f"--features={feature_set}",
+            "--all-targets",
+            "--workspace",
+            "--",
+            "-Dwarnings",
```
Comment (adamreichold) [inline noxfile.py:57]: To improve self-documentation of such scripts, I would prefer the long option forms, i.e. `--deny warnings`.

File: noxfile.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+
+@nox.session(name="fmt-rust", venv_backend="none")
+def fmt_rust(session: nox.Session):
+    session.run("cargo", "fmt", "--all", "--check", external=True)
+
+
+@nox.session(name="fmt-py")
+def fmt_py(session: nox.Session):
+    session.install("black==21.12b0")
```
Comment (adamreichold) [inline noxfile.py:43]: I think `black==22.1.0` is current but I am not sure whether bumping this is wanted or sensible to do in this PR.

Comment (davidhewitt) [reply noxfile.py:43]: I'm happy to do so. Will push as a separate commit 👍

**Pull request merged**